core/src/main/java/software/amazon/awssdk/core/util/UserAgentUtils.java

```diff
@@ -73,33 +73,33 @@ static String userAgent() {
         ua = ua
                 .replace("{platform}", "java")
                 .replace("{version}", VersionInfo.SDK_VERSION)
-                .replace("{os.name}", replaceSpaces(JavaSystemSetting.OS_NAME.getStringValue().orElse(null)))
-                .replace("{os.version}", replaceSpaces(JavaSystemSetting.OS_VERSION.getStringValue().orElse(null)))
-                .replace("{java.vm.name}", replaceSpaces(JavaSystemSetting.JAVA_VM_NAME.getStringValue().orElse(null)))
-                .replace("{java.vm.version}", replaceSpaces(JavaSystemSetting.JAVA_VM_VERSION.getStringValue().orElse(null)))
-                .replace("{java.version}", replaceSpaces(JavaSystemSetting.JAVA_VERSION.getStringValue().orElse(null)))
+                .replace("{os.name}", sanitizeInput(JavaSystemSetting.OS_NAME.getStringValue().orElse(null)))
+                .replace("{os.version}", sanitizeInput(JavaSystemSetting.OS_VERSION.getStringValue().orElse(null)))
+                .replace("{java.vm.name}", sanitizeInput(JavaSystemSetting.JAVA_VM_NAME.getStringValue().orElse(null)))
+                .replace("{java.vm.version}", sanitizeInput(JavaSystemSetting.JAVA_VM_VERSION.getStringValue().orElse(null)))
+                .replace("{java.version}", sanitizeInput(JavaSystemSetting.JAVA_VERSION.getStringValue().orElse(null)))
                 .replace("{additional.languages}", getAdditionalJvmLanguages());
 
         Optional<String> language = JavaSystemSetting.USER_LANGUAGE.getStringValue();
         Optional<String> region = JavaSystemSetting.USER_REGION.getStringValue();
 
         String languageAndRegion = "";
         if (language.isPresent() && region.isPresent()) {
-            languageAndRegion = " " + replaceSpaces(language.get()) + "_" + replaceSpaces(region.get());
+            languageAndRegion = " (" + sanitizeInput(language.get()) + "_" + sanitizeInput(region.get()) + ")";
         }
         ua = ua.replace("{language.and.region}", languageAndRegion);
 
         return ua;
     }
 
     /**
-     * Replace any spaces in the input with underscores.
+     * Replace any spaces, parentheses in the input with underscores.
      *
      * @param input the input
      * @return the input with spaces replaced by underscores
      */
-    private static String replaceSpaces(final String input) {
-        return input == null ? UNKNOWN : input.replace(' ', '_');
+    private static String sanitizeInput(String input) {
+        return input == null ? UNKNOWN : input.replaceAll("[() ]", "_");
```

Comment:
Should we use a whitelist instead of a blacklist? https://tools.ietf.org/html/rfc7230#section-3.2.6

Reply:
the whitelist seems to be very long and I don't think jvm, os name and versions are likely to have non-ascii characters. But I'll definitely add the rest of the disallowed characters:`"(),/:;<=>?@[\]{}`